Adding function sun_position_el (ported from SatelliteToolbox.jl)

spacenewb opened the pull request:
Added `sun_position_el` and associated tests

spacenewb committed:
Added sun_position_el function and its test

Computes sun position in local coordinates.  Ported from SatelliteToolbox.jl

spacenewb committed:
Update sun.jl

updated tests to include all algorithms with independent tolerances

spacenewb committed:
Update sun.jl

fixed independent tolerances for test of sun_position_el

ronisbr commented:
Hi @spacenewb!

Thanks!

I have some comments that I will provide as a review in the source-code.

Furthermore, I think we should not add those functionalities in SatelliteToolboxBase.jl. Otherwise, packages that will not need all those functionalities will need to load them. The correct approach is to add SatelliteToolboxTransformations.jl as dependency to this package. Can you please make this modification?

spacenewb committed:
Update SatelliteToolboxCelestialBodies.jl

Added SatelliteToolboxTransformations.jl as a dependency to this package

spacenewb committed:
Updated associated tests and algorithm error

The algorithms to compute the RA and DEC gave negative angular results. This is corrected to give only positive angular results.

spacenewb commented:
Sure!

It was tardy on my part. I was too lazy to check the source code in SatelliteToolboxTransformations.jl. I have corrected the duplicate import issues now. There was a minor bug in the `sun_position_el` function as well, which is now corrected. On my local system, SatelliteToolboxCelestialBodies.jl now passes all the tests.

spacenewb committed:
Update Project.toml

Manually added dependency and compatibility for SatelliteToolboxTransformations.jl

spacenewb committed:
Changed the algorithm for speed and robustness

The previous algorithms used were case based and less robust. The currently implemented algorithm (PSA+) is a fast approximated solution up to 50 arcsecs accuracy.

spacenewb committed:
Updated function docstring and minor comments

spacenewb commented:
I changed the algorithm used by the function to compute the topocentric sun position. The new algorithm is called PSA+. It can be found at [Updating the PSA sun position algorithm](https://repository.cyi.ac.cy/handle/CyI/1059).

ronisbr commented:
Hi @spacenewb!

I left some comments in the code review section, can you please take a look? The most important one is related the function return be linked to a flag. In this case, the code will be type unstable and thus very slow.

spacenewb commented:
> Hi @spacenewb!
> 
> I left some comments in the code review section, can you please take a look? The most important one is related the function return be linked to a flag. In this case, the code will be type unstable and thus very slow.

Hi @ronisbr 

I cannot see any pending requests. I don't know if this is my mistake. Could you please check if you have left a code review

ronisbr commented:
Yes! I think so, take a look here: https://github.com/JuliaSpace/SatelliteToolboxCelestialBodies.jl/pull/2/files

The main problem is related to the `flag`, because it will add type instability.

spacenewb committed:
Removed `flag` from `sun_position_el`  in favour of type stability

Updated associated tests as well.

spacenewb commented:
I have updated the function `sun_position_el` such that it returns all output variables without any `flag`. I also happened to stumble upon the need to use the radius of the Earth as a constant. Do you think it's wise to export it as a constant in the `SatelliteToolboxBase` and then use it here in `SatelliteToolboxCelestialBodies`?

spacenewb commented:
> Yes! I think so, take a look here: https://github.com/JuliaSpace/SatelliteToolboxCelestialBodies.jl/pull/2/files
> 
> The main problem is related to the `flag`, because it will add type instability.

I still cannot see any reviews. I am sorry because I am learning the platform. I must be doing something wrong.

I get this when I filter the PR list on the basis of reviews:


Reviews - Not reviewed by me:

![NotReviewedByMe](https://github.com/JuliaSpace/SatelliteToolboxCelestialBodies.jl/assets/9540219/359fa44b-d450-4d36-991f-4840b56f0341)


Reviews - Changes requested:

![ChangesRequested](https://github.com/JuliaSpace/SatelliteToolboxCelestialBodies.jl/assets/9540219/6bcbb101-3542-473d-8e92-1da1e9cd124b)


However, I have addressed the main issue about the `flag` in `sun_position_el`.

ronisbr commented:
Hum, very strange, I am seeing the reviews in the code. Do you mind if I merge this commit and make the modifications myself?

spacenewb commented:
Please feel free to merge this commit. I am sorry that I could not do the modifications myself.
